### PR TITLE
[ROCm] Optimize concat_mla_q for CDNA3 (MI300X) and CDNA4 (MI355X)

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary

Optimize the ROCm performance of `concat_mla_q` kernel introduced in #34917, with architecture-specific improvements for both CDNA3 (MI300X) and CDNA4 (MI355X).

### 1. CDNA3 / gfx942 (MI300X): Non-temporal store hints

Add `__builtin_nontemporal_store` to `st128_cs` and `st32_cs` on gfx942+. The `concat_mla_q` kernel is a pure memory copy (write-once, no reuse), so bypassing L2 cache on writes avoids cache pollution and improves effective HBM bandwidth.

**Benchmark on MI300X** (non-contiguous nope input, bf16, CUDAGraph):

| num_tokens | Before (µs) | After (µs) | Speedup |
|------------|-------------|------------|---------|
| 128        | 6.1         | 5.4        | **+11%** |
| 256        | 14.0        | 12.8       | **+9%** |
| 1024       | 58.3        | 49.3       | **+15%** |
| 2048       | 157.3       | 121.8      | **+23%** |
| 4096       | 324.9       | 311.2      | +4% |
| 8192       | 638.6       | 619.6      | +3% |

> **Design note:** Read-side loads intentionally remain normal (not non-temporal) because NT loads hurt latency at small/medium token counts (< 2048 tokens). Only write-side uses NT hints.

### 2. CDNA4 / gfx950 (MI350X/MI355X): 256-bit wide load/store path

Enable `VLLM_256B_PTX_ENABLED` for gfx950, which doubles the logical vector width from 128-bit (CDNA3) to 256-bit (CDNA4), matching NVIDIA B300's `v8.u32` PTX capability.

**ISA reality (verified on MI350X, ROCm 7.2):** The gfx950 ISA does not have `global_load_dwordx8` / `global_store_dwordx8` vector instructions. The compiler decomposes 256-bit accesses into **2× `global_load/store_dwordx4`** with adjacent offsets. This still provides a meaningful optimization:

For NOPE_DIM=512 with bf16:
- **CDNA3 (128-bit):** 2 vectorized loop iterations, each loading 1× `dwordx4`
- **CDNA4 (256-bit):** 1 loop iteration, loading 2× `dwordx4` with `offset:16` (**halves loop overhead**)

Benefits even without single-instruction 256-bit access:
- 2× fewer loop iterations (fewer branches, index calculations)
- Adjacent `dwordx4` pairs can be pipelined by the memory controller
- Structurally ready for future ROCm versions if vector `dwordx8` is added

The kernel automatically selects the 256-bit path on gfx950 through the existing `use_256b` constexpr branch in `ConcatMLAQKernel` — no code changes needed in the kernel itself.

## Files Changed

- **`csrc/cuda_vec_utils.cuh`** (main changes):
  - Extend `VLLM_256B_PTX_ENABLED` to detect `__gfx950__`
  - Add `VLLM_ROCM_USE_NT_HINTS` macro for gfx942+
  - Add ROCm gfx950 paths for `ld256`/`st256`/`ld256_cs`/`st256_cs`
  - Add non-temporal store to `st128_cs`/`st32_cs` for gfx942+

- **`csrc/concat_mla_q.cuh`** (minor):
  - Fix includes for ROCm (`hip_bf16.h` / `hip_fp16.h` instead of CUDA headers)

## Testing

- ✅ Compile-tested on gfx942 with ROCm 6.4.3
- ✅ Correctness verified against CPU reference (0 errors, 4.7M elements)
- ✅ Performance benchmarked on MI300X with multiple token counts
- ✅ **ISA verified on MI350X (gfx950)** with ROCm 7.2 — confirms 2× `dwordx4` decomposition and working `nt` store hints
- ✅ Non-temporal store `nt` flag confirmed in gfx950 ISA output

## How to reproduce

```bash
# Build vllm for ROCm
python setup.py develop

# Run benchmark
python benchmarks/kernels/bench_concat_mla_q.py

# Run correctness tests
pytest tests/kernels/test_concat_mla_q.py -v

# ISA verification (gfx950)
hipcc --offload-arch=gfx950 --save-temps -O3 test.cpp
grep "dwordx" test-hip-amdgcn-amd-amdhsa-gfx950.s
```

## No environment variables needed

All optimizations are selected at **compile time** based on the GPU target architecture (`--offload-arch=gfx942` or `gfx950`). No runtime flags or environment variables are required.

### CODE REVIEW

File: csrc/cuda_vec_utils.cuh [hunk 1]
```diff
@@ -253,6 +300,12 @@ __forceinline__ __device__ void st128_cs(int4* addr, int4 val) {
 #ifndef USE_ROCM
   asm volatile("st.global.cs.v4.u32 [%0], {%1,%2,%3,%4};" ::"l"(addr),
                "r"(val.x), "r"(val.y), "r"(val.z), "r"(val.w));
+#elif VLLM_ROCM_USE_NT_HINTS
+  int* dst = reinterpret_cast<int*>(addr);
+  __builtin_nontemporal_store(val.x, dst);
+  __builtin_nontemporal_store(val.y, dst + 1);
+  __builtin_nontemporal_store(val.z, dst + 2);
+  __builtin_nontemporal_store(val.w, dst + 3);
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:344]: Verified on MI300X — the compiler auto-coalesces the four scalar `__builtin_nontemporal_store` into a **single** `global_store_dwordx4 ... nt` instruction.

Actual ISA from `hipcc -O3 --offload-arch=gfx942 --save-temps` for `ConcatMLAQKernel<bf16, 512>`:

```asm
global_load_dwordx4  v[0:3],   v[12:13], off
global_load_dwordx4  v[4:7],   v[12:13], off offset:512
global_store_dwordx4 v[10:11], v[0:3],   off nt              ; single 128-bit NT store
global_store_dwordx4 v[10:11], v[4:7],   off offset:512 nt   ; single 128-bit NT store
global_load_dword    v12,      v[8:9],   off
global_store_dword   v[0:1],   v12,      off offset:1024 nt  ; 32-bit NT store
```

No de-vectorization occurs — the compiler generates the same instruction count as the plain path, just with the `nt` flag added. We also tested inline ASM with `global_store_dwordx4` + `v4i32` vector register constraints, but hipcc produces `invalid operand for instruction` errors in multi-template contexts. The builtin approach is both correct and portable.

File: csrc/cuda_vec_utils.cuh [hunk 2]
```diff
@@ -185,28 +207,47 @@ __device__ __forceinline__ void st128(T& val, T* ptr) {
...
 __forceinline__ __device__ u32x8_t ld256_cs(const u32x8_t* addr) {
 #if VLLM_256B_PTX_ENABLED
   u32x8_t val;
+  #if defined(USE_ROCM) && defined(__gfx950__)
+  // gfx950 (CDNA4): use global_load_dwordx8 with slc flag for streaming
+  const uint32_t* src = reinterpret_cast<const uint32_t*>(addr);
+  for (int i = 0; i < 8; i++) {
+    val.d[i] = __builtin_nontemporal_load(
+        reinterpret_cast<const int*>(src) + i);
+  }
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:253]: Good point — the gfx950 path is not yet hardware-tested. However, based on ISA verification on gfx942, the compiler does coalesce sequential `__builtin_nontemporal_load/store` calls into vectorized instructions at `-O3`.

For the gfx950 256-bit path specifically: we expect the compiler to coalesce 8 × `__builtin_nontemporal_load(int)` into `global_load_dwordx4` pairs (or ideally `global_load_dwordx8` if the compiler recognizes the pattern). Since we cannot verify the gfx950 ISA output without MI355X hardware or a ROCm toolchain with gfx950 support, I will add a TODO comment noting that this should be verified when gfx950 hardware becomes available.

If the compiler does not coalesce to `dwordx8` on gfx950, we can switch to inline ASM at that point. The `__builtin_nontemporal` approach was chosen because inline ASM with vector register constraints (`v4i32`/`v8i32`) causes `invalid operand for instruction` errors in hipcc when used across multiple template instantiations.

File: csrc/cuda_vec_utils.cuh [hunk 3]
```diff
@@ -185,28 +207,47 @@ __device__ __forceinline__ void st128(T& val, T* ptr) {
...
 
 __forceinline__ __device__ void st256_cs(u32x8_t* addr, u32x8_t val) {
 #if VLLM_256B_PTX_ENABLED
+  #if defined(USE_ROCM) && defined(__gfx950__)
+  // gfx950 (CDNA4): use non-temporal store for streaming write
+  int* dst = reinterpret_cast<int*>(addr);
+  for (int i = 0; i < 8; i++) {
+    __builtin_nontemporal_store(
+        static_cast<int>(val.d[i]), dst + i);
+  }
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:278]: Same reasoning as above — on gfx942, we verified that `__builtin_nontemporal_store` is coalesced into `global_store_dwordx4 nt` by the compiler. The gfx950 path should similarly benefit from compiler vectorization.

Will add a TODO comment for gfx950 ISA verification once hardware is available. If `dwordx8` coalescing does not happen automatically, inline ASM can be added as a follow-up.

File: csrc/cuda_vec_utils.cuh [hunk 4]
```diff
@@ -133,28 +143,52 @@ struct alignas(VecTraits<use_256b>::ARCH_MAX_VEC_SIZE) PackedVec {
...
+  #if defined(USE_ROCM) && defined(__gfx950__)
+  uint32_t* dst = reinterpret_cast<uint32_t*>(ptr);
+  dst[0] = val.d[0];
+  dst[1] = val.d[1];
+  dst[2] = val.d[2];
+  dst[3] = val.d[3];
+  dst[4] = val.d[4];
+  dst[5] = val.d[5];
+  dst[6] = val.d[6];
+  dst[7] = val.d[7];
```
Comment (LopezCastroRoberto) [inline csrc/cuda_vec_utils.cuh:194]: Maybe it’s also worth leaving a TODO to double-check, once the MI355X hardware is available, that the compiler automatically translates this into the corresponding `global_store_dwordx8` instruction, and not two sequential `global_store_dwordx4`?

File: csrc/cuda_vec_utils.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (LopezCastroRoberto) [reply csrc/cuda_vec_utils.cuh:253]: > we expect the compiler to coalesce 8 × __builtin_nontemporal_load(int) into global_load_dwordx4 pairs 

Just out of curiosity: any reason why this is not automatically translated to one `global_load_dwordx8` if the hardware supports it?

**Review by LopezCastroRoberto — COMMENTED** [review]
Hi @andyluo7,

I’m the author of the two PR optimizations referenced here. I was a bit conservative with the AMD path since I’m not very familiar with the low-level details there. However, as I plan to add more optimizations similar to this one in the future that will continue to affect ROCm, I thought it would be nice to connect.

I’ve left a few comments and curiosity questions below. Thanks!

File: csrc/cuda_vec_utils.cuh (hunk 4 above)
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:194]: Good suggestion — I'll add a TODO comment in the code.

To verify on MI355X once hardware is available:
```bash
hipcc -O3 --offload-arch=gfx950 --save-temps concat_mla_q.cu
# Check the .s file for:
#   global_store_dwordx8  (single 256-bit NT store)
# vs:
#   global_store_dwordx4 × 2  (two 128-bit NT stores)
```

If the compiler doesn't emit `dwordx8`, we can fall back to inline ASM with `global_store_dwordx8` directly. Pushing a commit with the TODO now.

File: csrc/cuda_vec_utils.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:253]: Great question! The reason is **hardware ISA generation differences**:

- **gfx942 (CDNA3 / MI300X)**: The maximum vector load width is 128-bit (`global_load_dwordx4`). There is no `global_load_dwordx8` instruction in the gfx942 ISA. So 8× scalar loads → 2× `global_load_dwordx4` is the best the compiler can do.

- **gfx950 (CDNA4 / MI355X)**: The ISA **does** include `global_load_dwordx8` (256-bit vector load). We expect the compiler to coalesce 8× scalar loads into a single `global_load_dwordx8` on this target — but this needs hardware verification (added TODO).

This 128-bit vs 256-bit vector width is also the main reason MI300X peaks at ~4.5x speedup vs B300's 11.8x for this kernel. MI355X should close that gap significantly.

Reference: [AMD CDNA4 ISA (gfx950)](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-cdna4-instruction-set-architecture.pdf) — see `GLOBAL_LOAD_DWORDX8` instruction.

File: csrc/cuda_vec_utils.cuh (hunk 4 above)
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:194]: Update: verified on MI350X (gfx950, ROCm 7.2)!

The compiler generates **2× `global_store_dwordx4`**, not a single `global_store_dwordx8`. The gfx950 ISA in ROCm 7.2 / LLVM 22.0 does not include vector `dwordx8` instructions at all (`llvm-mc` rejects them as invalid).

The 256-bit path still halves loop iterations, so the optimization is valid — just not via the single-instruction path we hoped for. Will add a TODO comment noting this for future ROCm versions.

File: csrc/cuda_vec_utils.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:253]: Now we have the answer from hardware testing! 🔬

The reason is simple: **`global_load/store_dwordx8` does not exist in the gfx950 ISA** (as of ROCm 7.2 / AMD LLVM 22.0). The maximum vector load width is still 128-bit (`dwordx4`), same as gfx942.

Only `s_buffer_load_dwordx8` (scalar buffer load) exists — there are no 256-bit *vector* load/store instructions.

So the compiler is doing the best it can: decomposing 256-bit into 2× `dwordx4` with adjacent offsets (`off` and `off offset:16`). It's possible a future ROCm release could add vector dwordx8 support, but for now this is a hardware/ISA limitation, not a compiler limitation.

File: csrc/cuda_vec_utils.cuh (hunk 4 above)
Comment (LopezCastroRoberto) [reply csrc/cuda_vec_utils.cuh:194]: Thanks for looking into this! Would it be better to disable the 256b memory instructions for now, then? Maybe with an assert "Not supported yet"?

Two consecutive 128b operations issued in different cycles would lead to uncoalesced memory accesses within a warp.

Comment (andyluo7) [reply csrc/cuda_vec_utils.cuh:194]: Good call — you're right that two consecutive 128b ops in different cycles would lead to uncoalesced accesses. I've disabled the gfx950 256-bit path for now (commented out with explanation).

The gfx942 (MI300X) non-temporal hints path remains unchanged — that one is hardware-verified and the compiler auto-coalesces into single `global_store_dwordx4 nt` instructions.

Also rebased on upstream/main to resolve the merge conflict.

When a future ROCm version adds native `global_load/store_dwordx8`, we can re-enable the 256-bit path with a simple uncomment + re-verify.

**Comment by andyluo7** [comment]
Rebased onto latest main — merge conflict resolved. All previous review feedback has been addressed (gfx950 256-bit path disabled per @LopezCastroRoberto's suggestion).

@LopezCastroRoberto Would you mind taking another look when you get a chance? The only remaining commit is the MI300X non-temporal store optimization (hardware-verified, +11-23% speedup). Ready for full CI when a reviewer adds the `ready` label. Thanks!

**Review by LopezCastroRoberto — APPROVED** [review]
Thanks for the updates, @andyluo7!

Overall, this looks good to me. I’m still unsure whether it would be better to remove the ROCm path from the 256b instructions for now and just let the assert trigger and show the message, rather than keeping the current branch which will never be reached, e.g.:

https://github.com/vllm-project/vllm/pull/36743/changes#diff-2c10794e2bf8cf6e7e057874836793c4e37dbc105164cc3a6a73f8f7d83f3c84R155

I give my approval for now, waiting to see what other reviewers think. Thanks!

File: csrc/cuda_vec_utils.cuh [hunk 5]
```diff
@@ -185,29 +232,58 @@ __device__ __forceinline__ void st128(T& val, T* ptr) {
...
+  }
+  #else
   asm volatile(
       "st.global.cs.v8.u32 [%0], {%1,%2,%3,%4,%5,%6,%7,%8};" ::"l"(addr),
       "r"(val.d[0]), "r"(val.d[1]), "r"(val.d[2]), "r"(val.d[3]), "r"(val.d[4]),
       "r"(val.d[5]), "r"(val.d[6]), "r"(val.d[7]));
+  #endif
 #else
-  assert(false && "st256_cs requires SM100+ with CUDA 12.9+");
+  assert(false && "st256_cs requires SM100+ with CUDA 12.9+ or ROCm gfx950+");
```
Comment (LopezCastroRoberto) [inline csrc/cuda_vec_utils.cuh]: nit: remove/improve "or ROCm gfx950+" message in all the `asserts`, since 256-bit path is disabled for now on gfx950

**Comment by andyluo7** [comment]
## ✅ Tested on MI355X (gfx950) — All Tests Pass

Tested on **8x AMD Instinct MI355X** (gfx950, ROCm 7.0.1) using the latest commit.

### Correctness
All **31/31** tests pass:
- `test_concat_mla_q_contiguous` (10/10) ✅
- `test_concat_mla_q_transposed_nope` (8/8) ✅
- `test_concat_mla_q_split_rope` (10/10) ✅
- `test_concat_mla_q_zero_tokens` (1/1) ✅
- `test_concat_mla_q_values_preserved` (2/2) ✅

### Kernel Performance (`bench_concat_mla_q.py`, transposed non-contiguous nope inputs)

| num_tokens | torch.cat (µs) | concat_mla_q (µs) | Baseline v0.18.0 (µs) | vs torch.cat | vs Baseline |
|---:|---:|---:|---:|---:|---:|
| 64 | 11.21 | 2.72 | 3.39 | **4.1x** | **+25%** |
| 128 | 19.98 | 4.04 | 5.01 | **4.9x** | **+24%** |
| 256 | 35.93 | 11.56 | 11.91 | 3.1x | +3% |
| 512 | 63.93 | 22.50 | 22.55 | 2.8x | ~0% |
| 1024 | 132.14 | 44.10 | 47.28 | 3.0x | **+7%** |
| 2048 | 254.44 | 88.86 | 106.94 | 2.9x | **+20%** |
| 4096 | 498.73 | 215.01 | 229.80 | 2.3x | +7% |
| 8192 | 999.40 | 423.60 | 440.69 | 2.4x | +4% |

**Key observations:**
- Non-temporal store hints deliver **+20-25% improvement** at decode-relevant batch sizes (64-128 tokens) and large batches (2048 tokens) on gfx950
- The 256-bit path is correctly disabled for gfx950 in the latest commit (no native `dwordx8` in ISA) — only NT store hints are active
- No regressions at any token count

**Test environment:** MI355X (gfx950:sramecc+:xnack-), ROCm 7.0.1, vLLM built from PR branch
